File: agent/sessions/conversation.py
"""Per-session conversation store backed by a HistoryLog.

Each ConversationStore owns the raw history, checkpoint, and archive for
one session, stored under ``memory/sessions/<id>/``. The shared
MemoryStore retains the global layers (memory, user, episodes, versions).
"""
from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

from agent.memory_history import HistoryLog

if TYPE_CHECKING:
    from agent.memory import MemoryStore


class ConversationStore:
    """Scoped conversation persistence for a single session."""

    def __init__(self, session_dir: Path) -> None:
        self.session_dir = session_dir
        self.session_dir.mkdir(parents=True, exist_ok=True)
        self.history_file = session_dir / "history.jsonl"
        self.checkpoint_file = session_dir / "_checkpoint.json"
        self.history_log = HistoryLog(session_dir, self.history_file)

    # ── history ─────────────────────────────────────────────────────
    def append_history(
        self, role: str, content: object, *, extra: dict | None = None
    ) -> None:
        import json as _json
        from datetime import datetime, timedelta, timezone

        _UTC8 = timezone(timedelta(hours=8))
        row: dict = {
            "ts": datetime.now(_UTC8).isoformat(timespec="seconds"),
            "role": role,
            "content": content if isinstance(content, str) else _json.dumps(content, ensure_ascii=False),
        }
        if extra:
            for k, v in extra.items():
                if k not in row:
                    row[k] = v
        self.history_log.append(row)

    def load_unarchived_history(self) -> list[dict]:
        out: list[dict] = []
        active = self.history_log.load_active_rows()
        hidden = {
            str(r.get("turn_id"))
            for r in active
            if isinstance(r.get("turn_id"), str)
            and (r.get("hidden") is True or r.get("schedulerHidden") is True)
        }
        for row in active:
            if "role" not in row or "content" not in row:
                continue
            if row.get("type") == "model_call":
                continue
            if str(row.get("turn_id") or "") in hidden:
                continue
            item: dict = {"role": row["role"], "content": row["content"]}
            if isinstance(row.get("turn_id"), str):
                item["turn_id"] = row["turn_id"]
            out.append(item)
        return out

    def load_unarchived_turn_ids(self) -> list[str]:
        ids: list[str] = []
        seen: set[str] = set()
        for row in self.load_unarchived_history():
            if "role" in row and "content" in row:
                tid = row.get("turn_id")
                if isinstance(tid, str) and tid and tid not in seen:
                    seen.add(tid)
                    ids.append(tid)
        return ids

    def append_compact_marker(
        self, active_history: list[dict] | None = None
    ) -> None:
        if active_history is None:
            row = {"ts": "", "type": "compact_event"}
            self.history_log.append(row)
            return
        self.history_log.compact(active_history)

    def stats(self) -> dict:
        return self.history_log.stats()

    # ── checkpoint ──────────────────────────────────────────────────
    def write_checkpoint(self, history: list[dict]) -> None:
        import json as _json
        from datetime import datetime, timedelta, timezone

        _UTC8 = timezone(timedelta(hours=8))
        payload = {
            "ts": datetime.now(_UTC8).isoformat(timespec="seconds"),
            "history": history,
        }
        tmp = self.checkpoint_file.with_suffix(".json.tmp")
        tmp.write_text(_json.dumps(payload, ensure_ascii=False), encoding="utf-8")
        tmp.replace(self.checkpoint_file)

    def read_checkpoint(self) -> list[dict] | None:
        import json as _json

        if not self.checkpoint_file.exists():
            return None
        try:
            data = _json.loads(self.checkpoint_file.read_text(encoding="utf-8"))
        except (_json.JSONDecodeError, OSError):
            return None
        history = data.get("history") if isinstance(data, dict) else None
        return history if isinstance(history, list) else None

    def clear_checkpoint(self) -> None:
        self.checkpoint_file.unlink(missing_ok=True)


class SessionMemoryStore:
    """Conversation-scoped history with shared long-term memory layers."""

    def __init__(self, shared_memory: MemoryStore, conversation: ConversationStore) -> None:
        self.shared_memory = shared_memory
        self.conversation = conversation
        self.history_file = conversation.history_file
        self.checkpoint_file = conversation.checkpoint_file

    def append_history(self, role: str, content: object, *, extra: dict | None = None) -> None:
        self.conversation.append_history(role, content, extra=extra)

    def load_unarchived_history(self) -> list[dict]:
        return self.conversation.load_unarchived_history()

    def load_unarchived_turn_ids(self) -> list[str]:
        return self.conversation.load_unarchived_turn_ids()

    def append_compact_marker(self, active_history: list[dict] | None = None) -> None:
        self.conversation.append_compact_marker(active_history)

    def history_stats(self) -> dict:
        return self.conversation.stats()

    def write_checkpoint(self, history: list[dict]) -> None:
        self.conversation.write_checkpoint(history)

    def read_checkpoint(self) -> list[dict] | None:
        return self.conversation.read_checkpoint()

    def clear_checkpoint(self) -> None:
        self.conversation.clear_checkpoint()

    def __getattr__(self, name: str):
        return getattr(self.shared_memory, name)


class ProjectSessionMemoryStore(SessionMemoryStore):
    """Session history scoped to a project, with project AGENTS.md memory."""

    def __init__(
        self,
        shared_memory: MemoryStore,
        conversation: ConversationStore,
        project_store,
        project_id: str,
    ) -> None:
        super().__init__(shared_memory, conversation)
        self.project_store = project_store
        self.project_id = project_id

    def read_memory(self) -> str:
        return self.project_store.read_managed_memory(self.project_id)

    def write_memory(self, content: str) -> None:
        self.project_store.update_memory(self.project_id, content)

    def read_today_episode(self) -> str:
        return ""

    def append_episode(self, content: str) -> None:
        return None

    def write_user(self, content: str) -> None:
        return None

File: packages/core/src/sessions/conversation.ts
import { existsSync, mkdirSync, readFileSync, renameSync, rmSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'
import { HistoryLog } from '../memory/history'
import type { MemoryStore } from '../memory/store'
import { nowIsoUtc8 } from '../memory/time-utc8'

type Row = Record<string, unknown>

export class ConversationStore {
  readonly sessionDir: string
  readonly historyFile: string
  readonly checkpointFile: string
  readonly historyLog: HistoryLog

  constructor(sessionDir: string) {
    this.sessionDir = sessionDir
    mkdirSync(this.sessionDir, { recursive: true })
    this.historyFile = join(this.sessionDir, 'history.jsonl')
    this.checkpointFile = join(this.sessionDir, '_checkpoint.json')
    this.historyLog = new HistoryLog(this.sessionDir, this.historyFile)
  }

  appendHistory(role: string, content: unknown, opts?: { extra?: Row | null }): void {
    const row: Row = {
      ts: nowIsoUtc8(),
      role,
      content: typeof content === 'string' ? content : JSON.stringify(jsonSafe(content)),
    }
    if (opts?.extra) {
      for (const [key, value] of Object.entries(jsonSafe(opts.extra) as Row)) {
        if (!(key in row)) row[key] = value
      }
    }
    this.historyLog.append(row)
  }

  loadUnarchivedHistory(): Row[] {
    const out: Row[] = []
    const active = this.historyLog.loadActiveRows()
    const hidden = new Set<string>()
    for (const row of active) {
      if (typeof row.turn_id === 'string' && (row.hidden === true || row.schedulerHidden === true)) {
        hidden.add(row.turn_id)
      }
    }
    for (const row of active) {
      if (!('role' in row) || !('content' in row)) continue
      if (row.type === 'model_call') continue
      if (hidden.has(String(row.turn_id ?? ''))) continue
      const item: Row = { role: row.role, content: row.content }
      if (typeof row.turn_id === 'string') item.turn_id = row.turn_id
      if (Array.isArray(row.attachments)) item.attachments = row.attachments
      if (typeof row.displayContent === 'string') item.displayContent = row.displayContent
      out.push(item)
    }
    return out
  }

  loadUnarchivedTurnIds(): string[] {
    const ids: string[] = []
    const seen = new Set<string>()
    for (const row of this.loadUnarchivedHistory()) {
      const turnId = row.turn_id
      if (typeof turnId !== 'string' || !turnId || seen.has(turnId)) continue
      seen.add(turnId)
      ids.push(turnId)
    }
    return ids
  }

  appendCompactMarker(activeHistory?: Row[] | null): void {
    if (activeHistory === undefined || activeHistory === null) {
      this.historyLog.append({ ts: '', type: 'compact_event' })
      return
    }
    this.historyLog.compact(activeHistory)
  }

  stats(): Row {
    return this.historyLog.stats()
  }

  writeCheckpoint(history: Row[]): void {
    const payload = { ts: nowIsoUtc8(), history: jsonSafe(history) }
    const tmp = this.checkpointFile.replace(/\.json$/, '') + '.json.tmp'
    writeFileSync(tmp, JSON.stringify(payload), 'utf8')
    renameSync(tmp, this.checkpointFile)
  }

  readCheckpoint(): Row[] | null {
    if (!existsSync(this.checkpointFile)) return null
    try {
      const data = JSON.parse(readFileSync(this.checkpointFile, 'utf8'))
      const history = data && typeof data === 'object' && !Array.isArray(data) ? data.history : null
      return Array.isArray(history) ? history : null
    } catch {
      return null
    }
  }

  clearCheckpoint(): void {
    rmSync(this.checkpointFile, { force: true })
  }
}

export class SessionMemoryStore {
  readonly sharedMemory: MemoryStore
  readonly conversation: ConversationStore
  readonly memoryDir: string
  readonly historyFile: string
  readonly checkpointFile: string

  constructor(sharedMemory: MemoryStore, conversation: ConversationStore) {
    this.sharedMemory = sharedMemory
    this.conversation = conversation
    this.memoryDir = sharedMemory.memoryDir
    this.historyFile = conversation.historyFile
    this.checkpointFile = conversation.checkpointFile
  }

  appendHistory(role: string, content: unknown, opts?: { extra?: Row | null }): void {
    this.conversation.appendHistory(role, content, opts)
  }

  loadUnarchivedHistory(): Row[] {
    return this.conversation.loadUnarchivedHistory()
  }

  loadUnarchivedTurnIds(): string[] {
    return this.conversation.loadUnarchivedTurnIds()
  }

  appendCompactMarker(activeHistory?: Row[] | null): void {
    this.conversation.appendCompactMarker(activeHistory)
  }

  historyStats(): Row {
    return this.conversation.stats()
  }

  writeCheckpoint(history: Row[]): void {
    this.conversation.writeCheckpoint(history)
  }

  readCheckpoint(): Row[] | null {
    return this.conversation.readCheckpoint()
  }

  clearCheckpoint(): void {
    this.conversation.clearCheckpoint()
  }

  readMemory(): string { return this.sharedMemory.readMemory() }
  writeMemory(content: string): void { this.sharedMemory.writeMemory(content) }
  readTodayEpisode(): string { return this.sharedMemory.readTodayEpisode() }
  appendEpisode(content: string): void { this.sharedMemory.appendEpisode(content) }
  readUser(): string { return this.sharedMemory.readUser() }
  writeUser(content: string): void { this.sharedMemory.writeUser(content) }
}

export interface ProjectMemoryStoreLike {
  readManagedMemory(projectId: string): string
  updateMemory(projectId: string, content: string): void
}

export class ProjectSessionMemoryStore extends SessionMemoryStore {
  readonly projectStore: ProjectMemoryStoreLike
  readonly projectId: string

  constructor(
    sharedMemory: MemoryStore,
    conversation: ConversationStore,
    projectStore: ProjectMemoryStoreLike,
    projectId: string,
  ) {
    super(sharedMemory, conversation)
    this.projectStore = projectStore
    this.projectId = projectId
  }

  override readMemory(): string {
    return this.projectStore.readManagedMemory(this.projectId)
  }

  override writeMemory(content: string): void {
    this.projectStore.updateMemory(this.projectId, content)
  }

  override readTodayEpisode(): string {
    return ''
  }

  override appendEpisode(_content: string): void {
    return undefined
  }

  override writeUser(_content: string): void {
    return undefined
  }
}

function jsonSafe(obj: unknown): unknown {
  try {
    JSON.stringify(obj)
    return obj
  } catch {
    if (Array.isArray(obj)) return obj.map(jsonSafe)
    if (obj && typeof obj === 'object') {
      const out: Row = {}
      for (const [key, value] of Object.entries(obj as Row)) out[key] = jsonSafe(value)
      return out
    }
    return String(obj)
  }
}
